File: DailyChallenges/CPP/LeetCode_931.cpp
// 931. Minimum Falling Path Sum @ 2022/12/13

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        for (int i=1; i<m; ++i)
            for(int j = 0; j<n; ++j)
                matrix[i][j] +=  min(min((j==0)? INT_MAX: matrix[i-1][j-1], matrix[i-1][j]), (j==n-1)? INT_MAX: matrix[i-1][j+1]);
        return *min_element(matrix[m-1].begin(), matrix[m-1].end());
    }
};

File: DailyChallenges/Scala/LeetCode_931.scala
object Solution {
    def minFallingPathSum(matrix: Array[Array[Int]]): Int = {
        val m = matrix.length; val n = matrix(0).length
        for (i <- 1 until m; j <- 0 until n)
            matrix(i)(j) += math.min(math.min(if (j==0) Int.MaxValue else matrix(i-1)(j-1), matrix(i-1)(j)), if (j == n-1) Int.MaxValue else matrix(i-1)(j+1))
        matrix(m-1).min
    }
}
